apollo-router/src/plugins/authentication/subgraph.rs

```diff
@@ -0,0 +1,582 @@
+    tracing::info!(
+        monotonic_counter.apollo.router.operations.authentication.aws.sigv4 = 1u64,
+        authentication.aws.sigv4.failed = true,
+        subgraph.service.name = %subgraph_name,
+    );
+}
+
+pub(super) async fn make_signing_params(
+    config: &AuthConfig,
+    subgraph_name: &str,
+) -> Result<SigningParamsConfig, BoxError> {
+    match config {
+        AuthConfig::AWSSigV4(config) => {
+            let credentials_provider = config.get_credentials_provider().await;
+            if let Err(e) = credentials_provider.provide_credentials().await {
+                let error_subgraph_name = if subgraph_name == "all" {
+                    "all subgraphs".to_string()
+                } else {
+                    format!("{} subgraph", subgraph_name)
+                };
+                return Err(format!(
+                    "auth: {}: couldn't get credentials from provider: {}",
+                    error_subgraph_name, e,
+                )
+                .into());
+            }
+
+            Ok(SigningParamsConfig {
+                region: config.region(),
+                service_name: config.service_name(),
+                credentials_provider,
+            })
+        }
+    }
+}
+
+/// There are three possible cases
+/// https://github.com/awslabs/aws-sdk-rust/blob/9c3168dafa4fd8885ce4e1fd41cec55ce982a33c/sdk/aws-sigv4/src/http_request/sign.rs#L264C1-L271C6
+fn get_signing_settings(signing_params: &SigningParamsConfig) -> SigningSettings {
+    let mut settings = SigningSettings::default();
+    settings.payload_checksum_kind = match signing_params.service_name.as_str() {
+        "s3" | "vpc-lattice-svcs" => PayloadChecksumKind::XAmzSha256,
+        _ => PayloadChecksumKind::NoHeader,
+    };
+    settings
+}
+
+pub(super) struct SubgraphAuth {
+    pub(super) signing_params: SigningParams,
+}
+
+impl SubgraphAuth {
+    pub(super) fn subgraph_service(
+        &self,
+        name: &str,
+        service: crate::services::subgraph::BoxService,
+    ) -> crate::services::subgraph::BoxService {
+        if let Some(signing_params) = self.params_for_service(name) {
+            let name = name.to_string();
+            ServiceBuilder::new()
+            .checkpoint_async(move |mut req: SubgraphRequest| {
+                let signing_params = signing_params.clone();
+                let name = name.clone();
+                async move {
+                    let credentials = match signing_params.credentials_provider.provide_credentials().await {
+                        Ok(credentials) => credentials,
+                        Err(err) => {
+                            tracing::error!(
+                                "Failed to serialize GraphQL body for AWS SigV4 signing, skipping signing. Error: {}",
+                                err
+                            );
+                            increment_failure_counter(name.as_str());
+                            return Ok(ControlFlow::Continue(req));
```

Comment:
As discussed, `increment_failure_counter` should be followed with `Err`

Reply:
fixed in [89e0d0a](https://github.com/apollographql/router/pull/3365/commits/89e0d0a8067e0b4819d8ec2d185045fc2bb0e49e)